homeassistant/components/kodi/binary_sensor.py

```diff
@@ -0,0 +1,165 @@
+"""Sensor classes for kodi."""
+from __future__ import annotations
+
+import logging
+
+from jsonrpc_base.jsonrpc import TransportError
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DATA_CONNECTION, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Kodi sensors based on a config entry."""
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    connman = data[DATA_CONNECTION]
+    if (uid := config_entry.unique_id) is None:
+        uid = config_entry.entry_id
+
+    sensor_entities = [
+        KodiBinaryEntity(
+            "Screensaver",
+            "GUI.OnScreensaverActivated",
+            "GUI.OnScreensaverDeactivated",
+            "System.ScreenSaverActive",
+            uid,
+            connman,
+        ),
+        KodiBinaryEntity(
+            "Energy saving",
+            "GUI.OnDPMSActivated",
+            "GUI.OnDPMSDeactivated",
+            "System.DPMSActive",
+            uid,
+            connman,
+        ),
+    ]
+    async_add_entities(sensor_entities)
+
+
+class KodiBinaryEntity(BinarySensorEntity):
+    """Generic binary sensor entity for WebSocket callbacks."""
+
+    _attr_has_entity_name = True
+
+    def __init__(self, name, api_on, api_off, boolean, uid, connman):
+        """Initialize kodi binary sensor."""
+        self._is_on = False
+        self._attr_name = name
+        self._attr_unique_id = uid + "_" + api_on
+        self._api_on = api_on
+        self._api_off = api_off
+        self._kodi_boolean = boolean
+        self._connman = connman
+        self._kodi_uid = uid
+
+    @property
+    def is_on(self):
+        """Return the state of the sensor."""
+        return self._is_on
+
+    @property
+    def device_info(self):
```

Comment:
you can set `self._attr_device_info` from within `__init__()`

Reply:
✅